absl/types/any.h

```diff
@@ -94,23 +94,20 @@ namespace absl {
 
 namespace any_internal {
 
-// FastTypeId<Type>() evaluates at compile/link-time to a unique integer for the
-// passed in type. Their values are neither contiguous nor small, making them
-// unfit for using as an index into a vector, but a good match for keys into
-// maps or straight up comparisons.
-// Note that on 64-bit (unix) systems size_t is 64-bit while int is 32-bit and
-// the compiler will happily and quietly assign such a 64-bit value to a
-// 32-bit integer. While a client should never do that it SHOULD still be safe,
-// assuming the BSS segment doesn't span more than 4GiB.
+template <typename Type>
+struct TypeTag {
+  constexpr static char dummy_var = 0;
+};
+
+template <typename Type>
+constexpr char TypeTag<Type>::dummy_var;
+
+// FastTypeId<Type>() evaluates at compile/link-time to a unique pointer for the
+// passed in type. These are meant to be good match for keys into maps or straight
+// up comparisons.
 template<typename Type>
-inline size_t FastTypeId() {
-  static_assert(sizeof(char*) <= sizeof(size_t),
-                "ptr size too large for size_t");
-
-  // This static variable isn't actually used, only its address, so there are
-  // no concurrency issues.
-  static char dummy_var;
-  return reinterpret_cast<size_t>(&dummy_var);
+constexpr inline const void* FastTypeId() {
+  return &TypeTag<Type>::dummy_var;
```

Comment:
If you have a constexpr `std::addressof` here, you'll want that to avoid ADL on the template parameter `Type`.

Reply:
There is no ADL here since `dummy_var` is always of type `char`, so `std::addressof` does not need to be used. The type `Type` does not matter. Side note: We do not have a portable constexpr `std::addressof` as it cannot be implemented without intrinsics. It *might* be possible for us to implement per supported configuration, but that has not yet been investigated and I suspect will only work for some.

Reply:
@calabrese-google Ah, sorry my mistake. I just saw the template parameter and reacted.

FYI here is the detection libc++ uses for `__builtin_addressof`. 

```c++
#if !__has_builtin(__builtin_addressof) && _GNUC_VER < 700
# define _LIBCPP_HAS_NO_BUILTIN_ADDRESSOF
#endif```

Reply:
Thanks. We'll likely do this at some point.